submariner-operator/move-subctl-to-its-own-repo.md

```diff
@@ -0,0 +1,69 @@
+Operator dependencies don't allow it. It also makes Operator upgrades difficult.
+
+## Proposal
+
+The proposal is to move `subctl` out of the `submariner-operator` repository to its own repository. Everything specific to `subctl` will be
+moved to the new `subctl` repository. The `subctl` project will depend on the Operator (_e.g._ for the CRDs, RBAC etc. which are defined in
+the Operator). The `submariner-operator` repository will only contain artifacts which the Operator framework generates and/or needs.
+
+### Pros
+
+* Operator framework upgrades would be simpler.
+* It's logically correct to separate the Operator and the library as they are 2 different deliverables and provide different ways to
+  install Submariner.
+* Testing criteria are different. We don't need `subctl`-related GHAs for Operator and vice versa.
+* Development approach is different. One is a CLI and another is a framework.
+* Integration into `submariner-addon` might be simplified.
+
+### Cons
+
+* A new repository to maintain.
+* Splitting up the Operator and `subctl` will require care when implementing changes in either which affect the other.
+
+## Design details
+
+A new GitHub repository `subctl` will be created under the `submariner-io` organisation. The `submariner-operator` repository content will
+be copied to the new `subctl` repository, by pushing the git repository as-is, so that the full history is preserved. Non-`subctl`-related
+content will then be removed from the `subctl` repository, and non-Operator-related content will be removed from the `submariner-operator`
+repository. Once this is complete, there should be no duplicate code content in the two repositories, other than the build framework. The
+`subctl` repository will be set up with the same permissions, branch setup, and code ownership as the `submariner-operator` repository.
+The `subctl` repository will also follow [the standard Go project layout](https://github.com/golang-standards/project-layout).
+
+## Work items
+
+### On `subctl` repository
+
+* Create a new `subctl` repository and copy all the contents from `submariner-operator` repository.
+* Delete Operator specific code. In particular, `bundle`, `config`, `controllers`, `deploy`, `packagemanifests`, `package/*-operator*`,
+  `scripts/operator*`, `bundle.Dockerfile` directories/files.
+* Change module name in `go.mod` to `github.com/submariner-io/subctl` and add a dependency on `submariner-operator`.
+
+### On `submariner-operator` repository
+
+* Delete the `subctl`-specific code.
+
+### On `get.submariner.io` repository
+
+* Change URLs to download `subctl` binary from new `subctl` repository.
+
+### On `submariner-website` repository
+
+* Point links to new `subctl` repository wherever applicable.
+
+### Release process
+
+* Add a `subctl` phase in release automation after `submariner-operator` phase.
```

Comment:
`subctl`’s `go.mod` would depend on `submariner-operator`’s.

Reply:
Specifically regarding CRD changes, we’d have to add the field to the operator, either with a default value or allowing empty values, and then update subctl. If we want to make the field compulsory with no default, we’d have to do it in two steps, but that would cause issues on upgrades anyway.

Reply:
An alternative approach is to add an "api" repo which hosts the various types (and clients) and then have both rely on that one, solving potential cyclical future nightmares.
A benefit of this is of course other projects can also import and use just the "api" repo and not be forced to import any other repo to get it.

We can leave this as an exercise for later though, as it's tangentially related to this change but would like to hear opinions about it.

Reply:
> An alternative approach is to add an "api" repo which hosts the various types (and clients) and then have both rely on that one, solving potential cyclical future nightmares. A benefit of this is of course other projects can also import and use just the "api" repo and not be forced to import any other repo to get it.
> 
> We can leave this as an exercise for later though, as it's tangentially related to this change but would like to hear opinions about it.

We’ve done this in the past; there’s no need for a separate repo, the APIs can be published as separate Go modules. See https://github.com/submariner-io/submariner-operator/pull/1544 for example.